src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -131,32 +131,55 @@ DropMarkedShards(bool waitForCleanupLock)
 			continue;
 		}
 
-		ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
-															placement->placementId);
-		ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
-
-		ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
-							 INT64_FORMAT " on %s:%d after it was moved away",
-							 shardPlacement->placementId, shardPlacement->shardId,
-							 shardPlacement->nodeName, shardPlacement->nodePort)));
-
-		/* prepare sql query to execute to drop the shard */
-		StringInfo dropQuery = makeStringInfo();
-		char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
-		appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+		if (TryDropShard(placement))
+		{
+			removedShardCount++;
+		}
+	}
 
-		List *dropCommandList = list_make2("SET LOCAL lock_timeout TO '1s'",
-										   dropQuery->data);
+	return removedShardCount;
+}
 
-		/* remove the shard from the node and the placement information */
-		SendCommandListToWorkerInSingleTransaction(shardPlacement->nodeName,
-												   shardPlacement->nodePort,
-												   NULL, dropCommandList);
 
+/*
+ * TryDropShard tries to drop the given shard placement and returns
+ * true on success. On failure, this method swallows errors and emits them
+ * as WARNINGs.
+ */
+static bool
+TryDropShard(GroupShardPlacement *placement)
+{
+	ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
+														placement->placementId);
+	ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
+
+	ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
+						 INT64_FORMAT " on %s:%d after it was moved away",
+						 shardPlacement->placementId, shardPlacement->shardId,
+						 shardPlacement->nodeName, shardPlacement->nodePort)));
+
+	/* prepare sql query to execute to drop the shard */
+	StringInfo dropQuery = makeStringInfo();
+	char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
+	appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+
+	/*
+	 * We set a lock_timeout here so that if there are running queries on the
+	 * shards we won't get blocked more than 1s and fail.
+	 */
```

Comment:
Maybe another thing to add is: The lock timeout also avoids getting stuck in a distributed deadlock, since we might be holding pg_dist_placement locks while also taking locks on the shard placements, and this code interrupts the distributed deadlock detector.